File: server/schemas.py
"""Pydantic request/response models for the inference API."""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field


class PredictionRequest(BaseModel):
    """Request body for /api/predict endpoint."""

    image: str = Field(
        ...,
        description="Base64-encoded fundus image",
    )
    return_gradcam: bool = Field(
        default=False,
        description="Whether to generate and return Grad-CAM heatmap",
    )


class ProbabilityDistribution(BaseModel):
    """Per-class probability distribution."""

    no_dr: float = Field(..., alias="No DR")
    mild_npdr: float = Field(..., alias="Mild NPDR")
    moderate_npdr: float = Field(..., alias="Moderate NPDR")
    severe_npdr: float = Field(..., alias="Severe NPDR")
    proliferative_dr: float = Field(..., alias="Proliferative DR")

    model_config = {"populate_by_name": True}


class PredictionResult(BaseModel):
    """Classification prediction result."""

    class_index: int = Field(..., alias="class", description="Predicted class index (0-4)")
    label: str = Field(..., description="Human-readable severity label")
    confidence: float = Field(..., description="Confidence score for predicted class")
    probabilities: dict[str, float] = Field(
        ..., description="Per-class probability distribution"
    )

    model_config = {"populate_by_name": True}


class GradCAMResult(BaseModel):
    """Grad-CAM visualization result."""

    heatmap: str = Field(..., description="Base64-encoded heatmap overlay image")
    attention_regions: list[str] = Field(
        default_factory=list,
        description="Detected attention region descriptions",
    )


class InferenceMetadata(BaseModel):
    """Metadata about the inference request."""

    model: str = Field(..., description="Model name and version")
    inference_time_ms: float = Field(..., description="Inference latency in milliseconds")
    preprocessing_applied: str = Field(..., description="Preprocessing method used")


class PredictionResponse(BaseModel):
    """Full response from /api/predict endpoint."""

    prediction: PredictionResult
    gradcam: Optional[GradCAMResult] = None
    metadata: InferenceMetadata


class HealthResponse(BaseModel):
    """Response from /health endpoint."""

    status: str = "ok"
    model_loaded: bool = False
    model_name: Optional[str] = None

File: frontend/src/lib/schemas.ts
import { z } from "zod/v4";

export const predictionRequestSchema = z.object({
  image: z.string().min(1, "Image is required"),
  return_gradcam: z.boolean().default(false),
});

export type PredictionRequest = z.infer<typeof predictionRequestSchema>;

export const predictionResultSchema = z.object({
  class: z.number().int().min(0).max(4),
  label: z.string(),
  confidence: z.number().min(0).max(1),
  probabilities: z.record(z.string(), z.number()),
});

export type PredictionResult = z.infer<typeof predictionResultSchema>;

export const gradcamResultSchema = z.object({
  heatmap: z.string(),
  attention_regions: z.array(z.string()),
});

export type GradCAMResult = z.infer<typeof gradcamResultSchema>;

export const inferenceMetadataSchema = z.object({
  model: z.string(),
  inference_time_ms: z.number(),
  preprocessing_applied: z.string(),
});

export type InferenceMetadata = z.infer<typeof inferenceMetadataSchema>;

export const predictionResponseSchema = z.object({
  prediction: predictionResultSchema,
  gradcam: gradcamResultSchema.nullable().optional(),
  metadata: inferenceMetadataSchema,
});

export type PredictionResponse = z.infer<typeof predictionResponseSchema>;

export const DR_GRADES = [
  { grade: 0, label: "No DR", color: "#22c55e", description: "No visible retinopathy" },
  { grade: 1, label: "Mild NPDR", color: "#eab308", description: "Microaneurysms only" },
  { grade: 2, label: "Moderate NPDR", color: "#f97316", description: "More than just microaneurysms" },
  { grade: 3, label: "Severe NPDR", color: "#ef4444", description: "Extensive intraretinal hemorrhages" },
  { grade: 4, label: "Proliferative DR", color: "#7c3aed", description: "Neovascularization or vitreous hemorrhage" },
] as const;
